### PR TITLE
Skip TimeReport Printing If No Data

### PR DESCRIPTION
**Pull request by chenhao94**

### CODE REVIEW

File: src/timer/timer_enable.cc [hunk 1]
```diff
@@ -224,16 +224,20 @@ std::unique_ptr<TimerReport> TimerStatCollector::GetReport(TimerSection* section
     report->section_name_    = section->name_;
     report->timing_duration_ = last_merge_time_ - last_clear_time_;
     for (std::size_t i = 0; i < kTimerEntryBucketNum; ++i) {
+        const auto bucket_hits = entry.duration_buckets_[i].hits_;
+        report->has_data_ |= bool(bucket_hits);
```
Comment (zhouxindong) [inline src/timer/timer_enable.cc:228]: 好象规范推荐cast最好使用static_cast

**Review by zhouxindong — APPROVED** [review]

File: src/timer/timer_enable.cc (hunk 1 above)
Comment (chenhao94) [reply src/timer/timer_enable.cc:228]: 这个不是cast，是`operator bool()`，还是说你觉得`!!`看着更好？

Comment (xkszltl) [reply src/timer/timer_enable.cc:228]: bool没有别的东西的构造函数，你这里是implicit cast了一次，应该用static_cast

File: src/timer/timer_enable.cc [hunk 2]
```diff
@@ -225,16 +225,20 @@ std::unique_ptr<TimerReport> TimerStatCollector::GetReport(TimerSection* section
     report->section_name_    = section->name_;
     report->timing_duration_ = last_merge_time_ - last_clear_time_;
     for (std::size_t i = 0; i < kTimerEntryBucketNum; ++i) {
+        const auto bucket_hits = entry.duration_buckets_[i].hits_;
+        report->has_data_ |= static_cast<bool>(bucket_hits);
         report->report_buckets_.emplace_back(impl::TimerStatEntryBucket{
```
Comment (xkszltl) [inline src/timer/timer_enable.cc:230]: Can it be push_back({})?

File: src/timer/timer_enable.cc [hunk 3]
```diff
@@ -313,6 +317,9 @@ void TimerReport::PrintToLog(unsigned indent_level) const {
     using std::chrono::duration_cast;
     using std::chrono::nanoseconds;
     const std::string indent(4 * indent_level, ' ');
```
Comment (xkszltl) [inline src/timer/timer_enable.cc:319]: Better to allow a general custom prefix, not just space indentations.

Comment (xkszltl) [inline src/timer/timer_enable.cc:319]: include

File: src/timer/timer_enable.cc [hunk 4]
```diff
@@ -313,6 +317,9 @@ void TimerReport::PrintToLog(unsigned indent_level) const {
     using std::chrono::duration_cast;
     using std::chrono::nanoseconds;
```
Comment (xkszltl) [inline src/timer/timer_enable.cc:318]: include

Comment (xkszltl) [inline src/timer/timer_enable.cc:318]: literal recommended.

File: src/timer/timer_enable.cc [hunk 5]
```diff
@@ -313,6 +317,9 @@ void TimerReport::PrintToLog(unsigned indent_level) const {
...
     const std::string indent(4 * indent_level, ' ');
+    if (!has_data_) {
+        return;
+    }
     LOG(INFO) << indent << "Section '" << GetSectionName() << "':";
```
Comment (xkszltl) [inline src/timer/timer_enable.cc:323]: Double quote.

File: src/timer/timer_enable.cc [hunk 6]
```diff
@@ -313,6 +317,9 @@ void TimerReport::PrintToLog(unsigned indent_level) const {
...
     if (GetTotalHits() > 0) {
         LOG(INFO) << indent << "    hits    : " << GetTotalHits() << " times, freq: " << GetFreq() << " Hz";
```
Comment (xkszltl) [inline src/timer/timer_enable.cc:325]: `XXX times at XXX Hz` would be shorter and more informative as it describe the relationship in-between.

File: src/timer/timer_enable.cc [hunk 7]
```diff
@@ -225,16 +225,20 @@ std::unique_ptr<TimerReport> TimerStatCollector::GetReport(TimerSection* section
     report->section_name_    = section->name_;
     report->timing_duration_ = last_merge_time_ - last_clear_time_;
     for (std::size_t i = 0; i < kTimerEntryBucketNum; ++i) {
+        const auto bucket_hits = entry.duration_buckets_[i].hits_;
+        report->has_data_ |= static_cast<bool>(bucket_hits);
```
Comment (xkszltl) [inline src/timer/timer_enable.cc:229]: Can this be derived from bkt.size()?

Comment (chenhao94) [reply src/timer/timer_enable.cc:229]: No, because the buckets are pre-allocated even if no hits.

File: src/timer/timer_enable.cc (hunk 4 above)
Comment (chenhao94) [reply src/timer/timer_enable.cc:318]: How? Can we do something like `GetXXXNumber()ns`?

File: src/timer/timer_enable.cc (hunk 3 above)
Comment (chenhao94) [reply src/timer/timer_enable.cc:319]: Let's do that when we have clear needs on it. It is irrelavant to this PR.

File: src/timer/timer_enable.cc (hunk 4 above)
Comment (xkszltl) [reply src/timer/timer_enable.cc:320]: nvm

**Pull request merged**